### PR TITLE
Register scale subresource for Deployments and ReplicaSets.

### PR DESCRIPTION
**Pull request by madhusudancs**
Only the last commit here needs review.

cc @bgrant0607 @lavalamp

### CODE REVIEW

**Comment by bgrant0607** [comment]
Sorry, I was looking at an obsolete commit.

Why are tests still failing?

**Comment by bgrant0607** [comment]
pkg/apiserver tests fail.

The API machinery has been changing a lot, so it would be useful if @caesarxuchao or @lavalamp could skim over those bits.

**Comment by lavalamp** [comment]
@bgrant0607 all the action is in the PR this is based on top of :)

**Comment by madhusudancs** [comment]
@bgrant0607 the failing unit test is from the api installer PR. I have fixed that test, but will push it out along with other changes I am making to that PR. This PR can't merge without it anyway.

**Comment by bgrant0607** [comment]
LGTM, but not adding label since other PR hasn't merged yet.

File: pkg/apis/autoscaling/helpers.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import (
+	"fmt"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+)
+
+// ScaleFromDeployment returns a scale subresource for a deployment.
+func ScaleFromDeployment(deployment *extensions.Deployment) (*Scale, error) {
```
Comment (lavalamp) [inline pkg/apis/autoscaling/helpers.go]: Move both of these functions to the package where they're used. I want this package to be all about type definitions if at all possible.

File: pkg/master/master.go [hunk 2]
```diff
@@ -174,6 +174,7 @@ func resetMetrics(w http.ResponseWriter, req *http.Request) {
 
 func (m *Master) InstallAPIs(c *Config) {
 	apiGroupsInfo := []genericapiserver.APIGroupInfo{}
+	autoscalingGroupVersion := unversioned.GroupVersion{Group: "autoscaling", Version: "v1"}
```
Comment (lavalamp) [inline pkg/master/master.go]: move this closer to where it's used, and I'd recommend repeating it inside each if block.

File: pkg/master/master.go [hunk 3]
```diff
@@ -263,7 +268,7 @@ func (m *Master) InstallAPIs(c *Config) {
 		autoscalingGroupMeta := registered.GroupOrDie(autoscaling.GroupName)
 
 		// Hard code preferred group version to autoscaling/v1
-		autoscalingGroupMeta.GroupVersion = unversioned.GroupVersion{Group: "autoscaling", Version: "v1"}
+		autoscalingGroupMeta.GroupVersion = autoscalingGroupVersion
```
Comment (lavalamp) [inline pkg/master/master.go]: Undo this change-- the autoscaling group versions should not be connected to the version of the sub resources.

File: pkg/master/master.go [hunk 4]
```diff
@@ -707,9 +712,10 @@ func (m *Master) getExtensionResources(c *Config) map[string]rest.Storage {
 		deploymentStorage := deploymentetcd.NewStorage(restOptions("deployments"))
 		storage["deployments"] = deploymentStorage.Deployment
 		storage["deployments/status"] = deploymentStorage.Status
-		// TODO(madhusudancs): Install scale when Scale group issues are fixed (see issue #18528).
-		// storage["deployments/scale"] = deploymentStorage.Scale
 		storage["deployments/rollback"] = deploymentStorage.Rollback
+		if !m.ApiGroupVersionOverrides["autoscaling/v1"].Disable {
```
Comment (lavalamp) [inline pkg/master/master.go]: I think we want to register this whether or not the group is registered?

File: pkg/master/master.go [hunk 5]
```diff
@@ -727,6 +733,9 @@ func (m *Master) getExtensionResources(c *Config) map[string]rest.Storage {
 		replicaSetStorage := replicasetetcd.NewStorage(restOptions("replicasets"))
 		storage["replicasets"] = replicaSetStorage.ReplicaSet
 		storage["replicasets/status"] = replicaSetStorage.Status
+		if !m.ApiGroupVersionOverrides["autoscaling/v1"].Disable {
```
Comment (lavalamp) [inline pkg/master/master.go]: Same here?

File: pkg/registry/deployment/etcd/etcd_test.go [hunk 6]
```diff
@@ -180,73 +185,75 @@ func TestWatch(t *testing.T) {
...
+			Replicas: validDeployment.Spec.Replicas,
+		},
+		Status: autoscaling.ScaleStatus{
+			Replicas: validDeployment.Status.Replicas,
+			Selector: selector.String(),
+		},
+	}
+}
+
+var validScale = *validNewScale()
```
Comment (lavalamp) [inline pkg/registry/deployment/etcd/etcd_test.go]: no need for this variable.

File: pkg/registry/deployment/etcd/etcd_test.go [hunk 7]
```diff
@@ -180,73 +185,75 @@ func TestWatch(t *testing.T) {
...
+	storage, server := newStorage(t)
+	defer server.Terminate(t)
+
+	ctx := api.WithNamespace(api.NewContext(), namespace)
+	key := etcdtest.AddPrefix("/deployments/" + namespace + "/" + name)
+	if err := storage.Deployment.Storage.Set(ctx, key, &validDeployment, nil, 0); err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	expect := &validScale
```
Comment (lavalamp) [inline pkg/registry/deployment/etcd/etcd_test.go]: Just call the function directly.

File: pkg/registry/deployment/etcd/etcd_test.go [hunk 8]
```diff
@@ -180,73 +185,75 @@ func TestWatch(t *testing.T) {
...
-// 	obj, err := storage.Deployment.Get(ctx, name)
-// 	if err != nil {
-// 		t.Fatalf("unexpected error: %v", err)
-// 	}
-// 	deployment := obj.(*extensions.Deployment)
-// 	if deployment.Spec.Replicas != replicas {
-// 		t.Errorf("wrong replicas count expected: %d got: %d", replicas, deployment.Spec.Replicas)
-// 	}
-// }
+func validNewScale() *autoscaling.Scale {
```
Comment (lavalamp) [inline pkg/registry/deployment/etcd/etcd_test.go]: make this function either take in the deployment as a parameter, or output a constant thing without reference to the deployment. It's super confusing to do this construction via a stack of global variables.

File: pkg/registry/deployment/etcd/etcd_test.go (hunk 7 above)
Comment (lavalamp) [inline pkg/registry/deployment/etcd/etcd_test.go]: actually if you just hard code your expectation here, it would be much easier to verify that this is correct.

File: pkg/registry/replicaset/etcd/etcd_test.go [hunk 9]
```diff
@@ -236,3 +241,106 @@ func TestWatch(t *testing.T) {
...
+			Replicas: validReplicaSet.Spec.Replicas,
+		},
+		Status: autoscaling.ScaleStatus{
+			Replicas: validReplicaSet.Status.Replicas,
+			Selector: selector.String(),
+		},
+	}
+}
+
+var validScale = *validNewScale()
```
Comment (lavalamp) [inline pkg/registry/replicaset/etcd/etcd_test.go]: same comments on this test

File: pkg/apis/autoscaling/helpers.go (hunk 1 above)
Comment (bgrant0607) [inline pkg/apis/autoscaling/helpers.go]: +1

File: pkg/apis/autoscaling/helpers.go [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+	selector, err := unversioned.LabelSelectorAsSelector(deployment.Spec.Selector)
+	if err != nil {
+		return nil, fmt.Errorf("invalid label selector: %v", err)
+	}
+	return &Scale{
+		ObjectMeta: api.ObjectMeta{
```
Comment (bgrant0607) [inline pkg/apis/autoscaling/helpers.go]: Please add a TODO to create a variant of ObjectMeta that only contains these fields.

File: pkg/apis/autoscaling/helpers.go (hunk 1 above)
Comment (bgrant0607) [inline pkg/apis/autoscaling/helpers.go]: As for rationale: The main idea behind Scale is to provide a polymorphic interface with multiple implementations. The implementations need to live the the respective controllers.

File: pkg/registry/deployment/etcd/etcd.go [hunk 11]
```diff
@@ -184,53 +186,51 @@ type ScaleREST struct {
...
-// 		return nil, false, errors.NewConflict(extensions.Resource("deployments/scale"), scale.Name, err)
-// 	}
-// 	newScale, err := extensions.ScaleFromDeployment(deployment)
-// 	if err != nil {
-// 		return nil, false, errors.NewBadRequest(fmt.Sprintf("%v", err))
-// 	}
-// 	return newScale, false, nil
-// }
+// ScaleREST implements Patcher
+var _ = rest.Patcher(&ScaleREST{})
```
Comment (bgrant0607) [inline pkg/registry/deployment/etcd/etcd.go]: @lavalamp FWIW, I find it confusing that "Patcher" means the resource supports Get and Put. Do all resources supporting Get and Put also support Patch?

File: pkg/master/master.go (hunk 4 above)
Comment (madhusudancs) [inline pkg/master/master.go]: That, I think, will be problematic for two reasons:

1. We call Creater.New(gvk) to get the versionedObject for decoding in api_installer.go. Creater.New(gvk) which really is Scheme.New(gvk) checks if gvk is registered before creating the object, fails otherwise. This is causing the tests to fail now. We could probably override this behavior by always creating the versionedObject for subresources from SubresourceGroupVersion override. But then
2. We will be returning objects whose type will be registered nowhere.

File: pkg/registry/deployment/etcd/etcd_test.go (hunk 8 above)
Comment (madhusudancs) [inline pkg/registry/deployment/etcd/etcd_test.go]: Removed this function entirely.

File: pkg/registry/deployment/etcd/etcd_test.go (hunk 7 above)
Comment (madhusudancs) [inline pkg/registry/deployment/etcd/etcd_test.go]: Done.

File: pkg/registry/deployment/etcd/etcd_test.go (hunk 6 above)
Comment (madhusudancs) [inline pkg/registry/deployment/etcd/etcd_test.go]: Removed.

File: pkg/master/master.go (hunk 3 above)
Comment (madhusudancs) [inline pkg/master/master.go]: Done.

File: pkg/master/master.go (hunk 2 above)
Comment (madhusudancs) [inline pkg/master/master.go]: Done.

File: pkg/apis/autoscaling/helpers.go (hunk 1 above)
Comment (madhusudancs) [inline pkg/apis/autoscaling/helpers.go]: Done.

Makes sense. I have made these functions local to the packages (unexported) for now. I will probably have to export them in the future when we implement scale support in kubectl.

File: pkg/registry/replicaset/etcd/etcd_test.go (hunk 9 above)
Comment (madhusudancs) [inline pkg/registry/replicaset/etcd/etcd_test.go]: Done.

File: pkg/apis/autoscaling/helpers.go (hunk 10 above)
Comment (madhusudancs) [inline pkg/apis/autoscaling/helpers.go]: Done.

**Comment by madhusudancs** [comment]
@lavalamp @bgrant0607 Have addressed the comments. PTAL.

File: pkg/registry/deployment/etcd/etcd_test.go [hunk 12]
```diff
@@ -180,73 +185,69 @@ func TestWatch(t *testing.T) {
...
-// 		t.Errorf("wrong replicas count expected: %d got: %d", replicas, deployment.Spec.Replicas)
-// 	}
-// }
+func TestScaleGet(t *testing.T) {
+	storage, server := newStorage(t)
+	defer server.Terminate(t)
+
+	ctx := api.WithNamespace(api.NewContext(), namespace)
+	key := etcdtest.AddPrefix("/deployments/" + namespace + "/" + name)
+	if err := storage.Deployment.Storage.Set(ctx, key, &validDeployment, nil, 0); err != nil {
```
Comment (pwittrock) [inline pkg/registry/deployment/etcd/etcd_test.go]: consider using github.com/stretchr/testify/assert

assert.NoError(t, storage.Deployment.Storage.Set(ctx, key, &validDeployment, nil, 0))


Same for following error checks

File: pkg/registry/deployment/etcd/etcd_test.go [hunk 13]
```diff
@@ -180,73 +185,69 @@ func TestWatch(t *testing.T) {
...
+			Replicas: validDeployment.Status.Replicas,
+			Selector: selector.String(),
+		},
+	}
+	obj, err := storage.Scale.Get(ctx, name)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	scale := obj.(*autoscaling.Scale)
+	if e, a := expect, scale; !api.Semantic.DeepDerivative(e, a) {
```
Comment (pwittrock) [inline pkg/registry/deployment/etcd/etcd_test.go]: Is it necessary to copy the pointers expected, scale to e, a?

Also consider:
assert.True(t, api.Semantic.DeepDerivative(e, a), util.ObjectDiff(e, a))

File: pkg/registry/deployment/etcd/etcd.go (hunk 11 above)
Comment (pwittrock) [inline pkg/registry/deployment/etcd/etcd.go]: Not sure if this answers your question, but the comment on Patcher is `// Patcher is a storage object that supports both get and update.` and the interface embeds `Getter` and `Updater`.

File: pkg/registry/deployment/etcd/etcd.go [hunk 14]
```diff
@@ -184,53 +187,76 @@ type ScaleREST struct {
...
+	return scale, nil
+}
+
+func (r *ScaleREST) Update(ctx api.Context, obj runtime.Object) (runtime.Object, bool, error) {
+	if obj == nil {
+		return nil, false, errors.NewBadRequest(fmt.Sprintf("nil update passed to Scale"))
+	}
+	scale, ok := obj.(*autoscaling.Scale)
+	if !ok {
+		return nil, false, errors.NewBadRequest(fmt.Sprintf("wrong object passed to Scale update: %v", obj))
```
Comment (pwittrock) [inline pkg/registry/deployment/etcd/etcd.go]: For RollbackREST.Create the format string when the object type does not match is: "expected input object type to be DeploymentRollback, but %T"   Should these be consistent?

**Comment by eparis** [comment]
@k8s-bot test this, issue #22435

File: pkg/registry/deployment/etcd/etcd.go [hunk 15]
```diff
@@ -184,53 +187,76 @@ type ScaleREST struct {
...
+		return nil, false, errors.NewNotFound(autoscaling.Resource("deployments/scale"), scale.Name)
+	}
+	deployment.Spec.Replicas = scale.Spec.Replicas
+	deployment, err = (*r.registry).UpdateDeployment(ctx, deployment)
+	if err != nil {
+		return nil, false, err
+	}
+	newScale, err := scaleFromDeployment(deployment)
+	if err != nil {
+		return nil, false, errors.NewBadRequest(fmt.Sprintf("%v", err))
```
Comment (pwittrock) [inline pkg/registry/deployment/etcd/etcd.go]: Is throwing an error the correct thing to do in this case?  We have just successfully updated the deployment with the new scale, but can't convert the deployment back to a scale because the label selectors are invalid on the deployment.  Is the deployment corrupted if the label selector is invalid?

File: pkg/registry/replicaset/etcd/etcd.go [hunk 16]
```diff
@@ -110,3 +120,81 @@ func (r *StatusREST) New() runtime.Object {
...
+	rs, err := (*r.registry).GetReplicaSet(ctx, scale.Name)
+	if err != nil {
+		return nil, false, errors.NewNotFound(autoscaling.Resource("replicasets/scale"), scale.Name)
+	}
+	rs.Spec.Replicas = scale.Spec.Replicas
+	rs, err = (*r.registry).UpdateReplicaSet(ctx, rs)
+	if err != nil {
+		return nil, false, err
+	}
+	newScale, err := scaleFromReplicaSet(rs)
```
Comment (pwittrock) [inline pkg/registry/replicaset/etcd/etcd.go]: Same comment about what to do in this case

**Comment by pwittrock** [comment]
Apologies for hoping on this so late in the game.  Given we are planning on cutting the 1.2 branch tomorrow morning, disregard comments (or add TODOs) for non-blockers.

File: pkg/registry/deployment/etcd/etcd_test.go (hunk 13 above)
Comment (madhusudancs) [inline pkg/registry/deployment/etcd/etcd_test.go]: Removed the e, a assignment and made the variables match Go testing lingo.

File: pkg/registry/deployment/etcd/etcd.go (hunk 14 above)
Comment (madhusudancs) [inline pkg/registry/deployment/etcd/etcd.go]: Done.

**Comment by pwittrock** [comment]
Test failures looks non-flaky
>
Running tests for APIVersion: v1,extensions/v1beta1,metrics/v1alpha1 with etcdPrefix: registry
+++ [0303 10:07:24] Running tests without code coverage
# k8s.io/kubernetes/pkg/registry/deployment/etcd
_output/local/go/src/k8s.io/kubernetes/pkg/registry/deployment/etcd/etcd_test.go:20: imported and not used: "fmt"
# k8s.io/kubernetes/pkg/registry/replicaset/etcd
_output/local/go/src/k8s.io/kubernetes/pkg/registry/replicaset/etcd/etcd_test.go:247: missing return at end of function

File: pkg/registry/deployment/etcd/etcd.go (hunk 15 above)
Comment (madhusudancs) [inline pkg/registry/deployment/etcd/etcd.go]: I think this is the right thing to do. Deployment should be considered corrupted if we can't get the selector.

File: pkg/registry/replicaset/etcd/etcd.go (hunk 16 above)
Comment (madhusudancs) [inline pkg/registry/replicaset/etcd/etcd.go]: Same answer as above.

**Comment by madhusudancs** [comment]
@pwittrock I have addressed your comments except using assertions. PTAL.

File: pkg/registry/deployment/etcd/etcd.go (hunk 15 above)
Comment (pwittrock) [inline pkg/registry/deployment/etcd/etcd.go]: In that case, we could probably return a more informative result indicating that:
1. The scale was updated on the deployment
2. There are issues with fields on the persisted deployment

Consider something like `errors.NewInvalid("Scale updated, but persisted Deployment has invalid fields: %v", err)`

File: pkg/registry/deployment/etcd/etcd.go [hunk 17]
```diff
@@ -184,53 +187,76 @@ type ScaleREST struct {
...
+	scale, ok := obj.(*autoscaling.Scale)
+	if !ok {
+		return nil, false, errors.NewBadRequest(fmt.Sprintf("expected input object type to be Scale, but %T", obj))
+	}
+
+	if errs := asvalidation.ValidateScale(scale); len(errs) > 0 {
+		return nil, false, errors.NewInvalid(autoscaling.Kind("Scale"), scale.Name, errs)
+	}
+
+	deployment, err := (*r.registry).GetDeployment(ctx, scale.Name)
```
Comment (pwittrock) [inline pkg/registry/deployment/etcd/etcd.go]: Is `(*r.registry)` necessary, or can we simplify to `r.registry.GetDeployment`?

Comment (madhusudancs) [inline pkg/registry/deployment/etcd/etcd.go]: It is required. Note that r.registry is a pointer to an interface. We need an interface to make these client calls.

Comment (pwittrock) [inline pkg/registry/deployment/etcd/etcd.go]: I see.  For my own education, when is it the convention / beneficial to use a pointer to an interface instead of the interface directly (since the interface itself is a pair of pointers)?

File: pkg/registry/deployment/etcd/etcd.go [hunk 18]
```diff
@@ -184,53 +187,76 @@ type ScaleREST struct {
...
+	if err != nil {
+		return nil, false, errors.NewNotFound(autoscaling.Resource("deployments/scale"), scale.Name)
+	}
+	deployment.Spec.Replicas = scale.Spec.Replicas
```
Comment (bgrant0607) [inline pkg/registry/deployment/etcd/etcd.go]: Please also copy ResourceVersion.

File: pkg/registry/replicaset/etcd/etcd.go [hunk 19]
```diff
@@ -110,3 +120,81 @@ func (r *StatusREST) New() runtime.Object {
...
+
+	if errs := asvalidation.ValidateScale(scale); len(errs) > 0 {
+		return nil, false, errors.NewInvalid(autoscaling.Kind("Scale"), scale.Name, errs)
+	}
+
+	rs, err := (*r.registry).GetReplicaSet(ctx, scale.Name)
+	if err != nil {
+		return nil, false, errors.NewNotFound(autoscaling.Resource("replicasets/scale"), scale.Name)
+	}
+	rs.Spec.Replicas = scale.Spec.Replicas
```
Comment (bgrant0607) [inline pkg/registry/replicaset/etcd/etcd.go]: Please also copy ResourceVersion.

File: pkg/registry/deployment/etcd/etcd.go (hunk 17 above)
Comment (lavalamp) [inline pkg/registry/deployment/etcd/etcd.go]: It's basically always wrong to use a pointer to an interface.

Comment (liggitt) [inline pkg/registry/deployment/etcd/etcd.go]: > It's basically always wrong to use a pointer to an interface.

+1

File: pkg/master/master.go (hunk 4 above)
Comment (lavalamp) [inline pkg/master/master.go]: registered != turned on. You're checking here if it's in runtime config, which would be used for turning hpa on and off. I agree it won't work if it's not in api versions-- but that's not what you're checking. @nikhiljindal to confirm

File: pkg/master/master.go [hunk 20]
```diff
@@ -242,6 +243,10 @@ func (m *Master) InstallAPIs(c *Config) {
 			Scheme:                 api.Scheme,
 			ParameterCodec:         api.ParameterCodec,
 			NegotiatedSerializer:   api.Codecs,
+			SubresourceGroupVersionKind: map[string]unversioned.GroupVersionKind{
```
Comment (lavalamp) [inline pkg/master/master.go]: If you do need to turn something off when autoscaling isn't on, should omit this, too.

File: pkg/registry/deployment/etcd/etcd.go (hunk 11 above)
Comment (lavalamp) [inline pkg/registry/deployment/etcd/etcd.go]: I think the machinery that implements patch just needs something that can Get and Put. You don't need anything special to get patch working on your object, logically it's just an update.

File: pkg/registry/deployment/etcd/etcd.go (hunk 18 above)
Comment (lavalamp) [inline pkg/registry/deployment/etcd/etcd.go]: +1 very important

File: pkg/registry/deployment/etcd/etcd.go [hunk 21]
```diff
@@ -184,53 +187,76 @@ type ScaleREST struct {
...
+		return nil, false, errors.NewBadRequest(fmt.Sprintf("%v", err))
+	}
+	return newScale, false, nil
+}
+
+// scaleFromDeployment returns a scale subresource for a deployment.
+func scaleFromDeployment(deployment *extensions.Deployment) (*autoscaling.Scale, error) {
+	selector, err := unversioned.LabelSelectorAsSelector(deployment.Spec.Selector)
+	if err != nil {
+		return nil, fmt.Errorf("invalid label selector: %v", err)
```
Comment (lavalamp) [inline pkg/registry/deployment/etcd/etcd.go]: this error will be super confusing. Say something like "stored deployment object can't be represented in the form of a scale subresource because the label selector ('%v') can't be parsed: %v"

File: pkg/registry/deployment/etcd/etcd_test.go (hunk 12 above)
Comment (lavalamp) [inline pkg/registry/deployment/etcd/etcd_test.go]: @pwittrock hard to recommend that as general practice. Go authors believe everyone should just write if statements. I'm pretty neutral, but I definitely don't think it's worth changing when this is a critical path PR.

File: pkg/registry/deployment/etcd/etcd_test.go [hunk 22]
```diff
@@ -180,73 +184,69 @@ func TestWatch(t *testing.T) {
...
+			Replicas: validDeployment.Status.Replicas,
+			Selector: selector.String(),
+		},
+	}
+	obj, err := storage.Scale.Get(ctx, name)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	got := obj.(*autoscaling.Scale)
+	if !api.Semantic.DeepDerivative(want, got) {
```
Comment (lavalamp) [inline pkg/registry/deployment/etcd/etcd_test.go]: don't use DeepDerivative unless you understand it completely and it's what you want.

Check RV.

File: pkg/registry/deployment/etcd/etcd_test.go [hunk 23]
```diff
@@ -180,73 +184,69 @@ func TestWatch(t *testing.T) {
...
+	}
+	replicas := 12
+	update := autoscaling.Scale{
+		ObjectMeta: api.ObjectMeta{Name: name, Namespace: namespace},
+		Spec: autoscaling.ScaleSpec{
+			Replicas: replicas,
+		},
+	}
+
+	if _, _, err := storage.Scale.Update(ctx, &update); err != nil {
```
Comment (lavalamp) [inline pkg/registry/deployment/etcd/etcd_test.go]: Test that wrong resource version gives a conflict.

File: pkg/registry/deployment/etcd/etcd_test.go (hunk 12 above)
Comment (pwittrock) [inline pkg/registry/deployment/etcd/etcd_test.go]: SGTM.  Not worth work changing in the PR.  Do we have any project guidelines around this yet?

File: pkg/master/master.go (hunk 4 above)
Comment (bgrant0607) [inline pkg/master/master.go]: Does registration of that group need to be optional any longer?

File: pkg/registry/deployment/etcd/etcd_test.go (hunk 12 above)
Comment (bgrant0607) [inline pkg/registry/deployment/etcd/etcd_test.go]: That looks nice, but isn't widely used in the code yet, so I wouldn't block on that.

File: pkg/registry/deployment/etcd/etcd_test.go [hunk 24]
```diff
@@ -180,73 +184,69 @@ func TestWatch(t *testing.T) {
...
-// 	}
-// }
+func TestScaleGet(t *testing.T) {
+	storage, server := newStorage(t)
+	defer server.Terminate(t)
+
+	ctx := api.WithNamespace(api.NewContext(), namespace)
+	key := etcdtest.AddPrefix("/deployments/" + namespace + "/" + name)
+	if err := storage.Deployment.Storage.Set(ctx, key, &validDeployment, nil, 0); err != nil {
+		t.Fatalf("unexpected error: %v", err)
```
Comment (bgrant0607) [inline pkg/registry/deployment/etcd/etcd_test.go]: However, if you don't use assert, ensure you print enough identifying information in log statements, such as the involved object and what was being done to it.

File: pkg/registry/deployment/etcd/etcd_test.go (hunk 22 above)
Comment (madhusudancs) [inline pkg/registry/deployment/etcd/etcd_test.go]: I think I want DeepDerivative but have switched to DeepEqual by filling in the fields which were nil/empty in `want` previously. Just to be safe.

File: pkg/registry/deployment/etcd/etcd.go (hunk 21 above)
Comment (madhusudancs) [inline pkg/registry/deployment/etcd/etcd.go]: Done.

File: pkg/registry/deployment/etcd/etcd.go (hunk 15 above)
Comment (madhusudancs) [inline pkg/registry/deployment/etcd/etcd.go]: Done. See the comment below.

File: pkg/registry/deployment/etcd/etcd.go (hunk 18 above)
Comment (madhusudancs) [inline pkg/registry/deployment/etcd/etcd.go]: Done.

File: pkg/registry/deployment/etcd/etcd.go (hunk 17 above)
Comment (madhusudancs) [inline pkg/registry/deployment/etcd/etcd.go]: Fixed. Have removed the pointer to interface.

File: pkg/master/master.go (hunk 4 above)
Comment (madhusudancs) [inline pkg/master/master.go]: Did not know the difference between registered vs turned on. Thanks for the clarification. Haven't fixed this yet. Will fix it soon.

File: pkg/registry/deployment/etcd/etcd_test.go (hunk 23 above)
Comment (madhusudancs) [inline pkg/registry/deployment/etcd/etcd_test.go]: Done.

File: pkg/registry/replicaset/etcd/etcd.go (hunk 19 above)
Comment (madhusudancs) [inline pkg/registry/replicaset/etcd/etcd.go]: Done.

File: pkg/registry/deployment/etcd/etcd_test.go (hunk 24 above)
Comment (madhusudancs) [inline pkg/registry/deployment/etcd/etcd_test.go]: Added additional information for all the error messages. Hopefully it is better now.

**Comment by madhusudancs** [comment]
@lavalamp @bgrant0607 @pwittrock Have addressed the comments. PTAL.

File: pkg/master/master.go (hunk 20 above)
Comment (madhusudancs) [inline pkg/master/master.go]: Done.

File: pkg/master/master.go (hunk 4 above)
Comment (madhusudancs) [inline pkg/master/master.go]: Fixed.

File: pkg/master/master.go (hunk 5 above)
Comment (madhusudancs) [inline pkg/master/master.go]: Fixed.

**Comment by pwittrock** [comment]
Does deployment/etcd/etcd.go need to be updated to include `deployment.Spec.Replicas = scale.Spec.Replicas`?

**Comment by madhusudancs** [comment]
> Does deployment/etcd/etcd.go need to be updated to include deployment.Spec.Replicas = scale.Spec.Replicas?

@pwittrock Where? It already does here - https://github.com/kubernetes/kubernetes/pull/21966/files#diff-4a2fa0c575e5ef68947993d029c7a5faR227

**Comment by lavalamp** [comment]
LGTM

**Comment by liggitt** [comment]
needs a squash

**Comment by lavalamp** [comment]
Squash please, then add label

**Comment by pwittrock** [comment]
@madhusudancs was looking at a wrong version of the file.  nm

**Comment by pwittrock** [comment]
LGTM

**Comment by madhusudancs** [comment]
@lavalamp squashed. Applying LGTM label.
